f[eat|ix](select): support values added by ngValue

Narretz opened the pull request:
select elements with ngModel will now set ngModel to option values added by ngValue.
This allows setting values of any type without the use of ngOptions.

Interpolations inside attributes can only be strings, but the ngValue directive uses attrs.$set,
which does not follow any type restriction. Any $observe on the value attribute will therefore receive
the original value (result of ngValue expression). However, when a user selects an option, the browser
sets the select value to the actual option's value attribute, which is still always a string.
For that reason, when option are added by ngValue, we set the hashed value of the original value in
the value attribute and store the actual value in an extra map. When the select value changes, we
read access the actual value via the hashed select value.

Closes #9842
Closes #6297
- [ ] Test that it takes precedence over interpolated option / option text
- [x] Test for multiple selections
- [x] Fix failing Chrome tests
- [ ] Docs

Talking points:
- The implementation works reasonably well, although it's a bit ~~hacky~~special
- With ngValue, option values are hash values, not the real values (which makes sense, but might be confusing)
- ~~afaik angular 2 suffers from the same problem as ng1 right now~~

Narretz committed:
feat(select): support values of any type added with ngValue

select elements with ngModel will now set ngModel to option values added by ngValue.
This allows setting values of any type (not only strings) without the use of ngOptions.

Interpolations inside attributes can only be strings, but the ngValue directive uses attrs.$set,
which does not have any type restriction. Any $observe on the value attribute will therefore receive
the original value (result of ngValue expression). However, when a user selects an option, the browser
sets the select value to the actual option's value attribute, which is still always a string.
For that reason, when options are added by ngValue, we set the hashed value of the original value in
the value attribute and store the actual value in an extra map. When the select value changes, we
read access the actual value via the hashed select value.

Since we only use a hashed value for ngValue, we will have extra checks for the hashed values:
- when options are read, for both single and multiple select
- when options are written, for multiple select

I don't expect this to have a performance impact, but it should be kept in mind.

Closes #9842
Closes #6297

BREAKING CHANGE:

`<option>` elements added to `<select ng-model>` via `ngValue` now add their values in hash form, i.e.
`<option ng-value="myString">` becomes `<option ng-value="myString" value="string:myString">`.

This is done to support binding options with values of any type to selects.

This should rarely affect applications, as the values of options are usually not relevant to the
application logic, but it's possible that option values are checked in tests.

Narretz commented on test/ng/directive/selectSpec.js:
```diff
@@ -1224,5 +1236,202 @@ describe('select', function() {
+          scope.selected = 'UPDATEDVALUE';
+          scope.$digest();
+
+          // expect(element.find('option').eq(0).prop('selected')).toBe(true); not selected in Chrome?
```
I wanted to fix this, but I'm not sure it's necessary. I don't think browsers have to set *selected* when the value of the select changes. It's strange though that is set before changing the select value via model update.

petebacondarwin commented on src/ng/directive/select.js:
```diff
@@ -116,9 +120,26 @@ var SelectController =
   self.registerOption = function(optionScope, optionElement, optionAttrs, interpolateValueFn, interpolateTextFn) {
 
-    if (interpolateValueFn) {
+    if (interpolateValueFn === true) {
```
I am not keen on using `true` as a special flag here.
Perhaps we could just add yet another parameter to the `registerOption` method?

petebacondarwin replied:
Or perhaps we don't need to do that at all, we could just check whether `optionAttrs.ngValue` is defined?

petebacondarwin commented:
This test fails:

```js
      it('should work update the value as a result of changes to the options', function() {
        var A = { name: 'A'}, B = { name: 'B'}, C = { name: 'C'};
        scope.options = [A, B, C];
        scope.obj = {};
        compile(
          '<select ng-model="obj.value">' +
            '<option ng-repeat="option in options" ng-value="option">{{$index}}</option>' +
          '</select>'
        );

        var optionElements = element.find('option');
        expect(optionElements.length).toEqual(4);
        browserTrigger(optionElements.eq(0));

        optionElements = element.find('option');
        expect(optionElements.length).toEqual(3);
        expect(scope.obj.value).toBe(A);

        scope.options.shift();
        scope.$digest();

        optionElements = element.find('option');
        expect(optionElements.length).toEqual(3);
        expect(scope.obj.value).toBe(null);
      });
```

Narretz commented:
Have you tried if that works with regular interpolated options? Because I don't think it does

petebacondarwin commented:
It certainly works with `ngOptions`, which is what we are trying to emulate, right?

Narretz committed:
fix(select): handle model updates when options are manipulated

These rules follow ngOptions behavior:

- when an option that is currently selected, is removed or its value changes, the model
is set to null.
- when an an option is added or its value changes to match the currently selected model,
this option is selected.
- when an option is disabled, the model is set to null.
- when the model value changes to a value that matches a disabled option,
this option is selected (analogue to ngOptions)

Narretz commented:
@petebacondarwin I've updated the PR again with the changes we talked about (support for disabled options, and model updates when the options change / are removed / get disabled / get added).

petebacondarwin commented on src/ng/directive/select.js:
```diff
@@ -113,36 +131,128 @@ var SelectController =
+          handleMultipleChanges = false;
+          self.ngModelCtrl.$setViewValue(self.readValue());
+          if (renderAfter) self.ngModelCtrl.$render();
+        });
```
This pattern of running something once in the postDigest keeps appearing. I wonder if we should abstract it out into $scope?

petebacondarwin commented:
I think you might have nailed it but it is the kind of change that I want to spend some time thinking about corner cases...

petebacondarwin committed:
chore(benchpress): add benchmarks for select with ngValue

Narretz committed:
perf(select): don't prepend unknown option if already prepended

petebacondarwin commented:
I did author some of these commits and I am OK with that :-)

petebacondarwin commented on src/ng/directive/select.js:
```diff
@@ -14,10 +14,15 @@ var SelectController =
   var self = this,
-      optionsMap = new HashMap();
+      optionsMap = new HashMap(),
+      handleMultipleDestroy = false; // Flag to run an update to the model after selected options
```
no longer needed?

Narretz commented:
I'll change the tests and then add some docs. Should I mentiond that performance in IE is going to be very bad with 3000+ options?

petebacondarwin commented:
Why not

Narretz commented:
While fixing up the PR, I found a behavior in ngOptions and select that I find a bit confusing: You have a that select is multiple and the model contains two values, one of which is matched by an option, while the other isn't.
What happens is that the matching option gets selected, and the model keeps both values. Imo, I would expect no option to be selected, as the model isn't completely matched by the select options. Thoughts?

Narretz commented:
So in https://github.com/angular/angular.js/issues/13172 we decided that if *no* option at all can be selected, then the select should be invalid (required). 

Doesn't say anything about a partial mismatch, though. We could decide not to select anything when not all options are matched, but I'm not convinced that this is a good idea.

Theoretically, setting the model to match the available options is the developers responsibility, so if that happens it's your own fault. However, it's not very easy to check if a select's options will match the model. We could expose the `selectCtrl.hasOption` function and make it work with ngOptions, so that users could manually check if a specific $viewValue is matched by their select, see (https://github.com/angular/angular.js/issues/12915)

Narretz commented:
I'm also fine with making a decision on this later. As far as I can tell, the described behavior hasn't bothered anyone yet.
@petebacondarwin I need your perf commit as part of the commit that handles multiple selects correctly. Because otherwise, the current behavior doesn't work as expected. If the model has two values set, but only one is matched by options, then without the "perf" fix, the option will set the model to the one value only. Delaying the update after render handles this.

gkalpak commented:
:sparkles: :tada: :sparkles:
